### PR TITLE
Pr

### PR DESCRIPTION
**Pull request by pristavkin-ivan**

### CODE REVIEW

File: core/src/main/java/com/epam/esm/gift_certificate/exception/TagCreationException.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    public TagCreationException(String message, Throwable cause) {
+        super(message, cause);
+    }
+
+    public TagCreationException(Throwable cause) {
+        super(cause);
+    }
+
+    @Override
+    public String getMessage() {
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/exception/TagCreationException.java:21]: that purpose of this method?

File: core/src/main/java/com/epam/esm/gift_certificate/dao/api/GiftCertificateDao.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ */
+public interface GiftCertificateDao<T extends GiftCertificate> {
+
+    /**
+     * Retrieves list of GiftCertificates from database
+     * @param query sql query
+     * @param words list of searching words
+     * @return List of GiftCertificates
+     */
+    List<T> getAll(String query, List<String> words);
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/dao/api/GiftCertificateDao.java:22]: it is better to have one object for search parameters

File: core/src/main/java/com/epam/esm/gift_certificate/dao/impl/GiftCertificateDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    public void update(GiftCertificate giftCertificate) throws NoSuchCertificateException {
+        GiftCertificate modifyingGiftCertificate = get(giftCertificate.getId()).get();
+
+        updateLogic(modifyingGiftCertificate, giftCertificate);
+
+        try {
+            jdbcOperations.update(SqlQueries.UPDATE_GIFT_CERTIFICATE, modifyingGiftCertificate.getName()
+                    , modifyingGiftCertificate.getDescription(), modifyingGiftCertificate.getPrice()
+                    , modifyingGiftCertificate.getDuration(), modifyingGiftCertificate.getId());
+        } catch (DataAccessException exception) {
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/dao/impl/GiftCertificateDaoImpl.java:97]: what if we get a db exception not related to resource absence?

File: core/src/main/java/com/epam/esm/gift_certificate/dao/impl/GiftCertificateDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+                , resultSet.getString(SqlLabels.G_NAME)
+                , resultSet.getString(SqlLabels.G_DESCRIPTION)
+                , resultSet.getDouble(SqlLabels.G_PRICE)
+                , resultSet.getInt(SqlLabels.G_DURATION)
+                , resultSet.getTimestamp(SqlLabels.G_CREATE_DATE)
+                , resultSet.getTimestamp(SqlLabels.G_LAST_UPDATE_DATE)
+                , new ArrayList<>());
+    }
+
+    private void updateLogic(GiftCertificate modifyingGiftCertificate, GiftCertificate giftCertificate) {
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/dao/impl/GiftCertificateDaoImpl.java:136]: this logic should be placed at service layer

File: core/src/main/java/com/epam/esm/gift_certificate/dao/impl/GiftCertificateDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        if (giftCertificate.getDuration() != null) {
+            modifyingGiftCertificate.setDuration(giftCertificate.getDuration());
+        }
+    }
+
+    private PreparedStatement createPreparedStatement(Connection connection, String name, String description
+            , Double price, Integer duration) throws SQLException {
+
+        PreparedStatement ps = connection.prepareStatement(SqlQueries.INSERT_GIFT_CERTIFICATES, new String[]{"id"});
+        ps.setString(1, name);
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/dao/impl/GiftCertificateDaoImpl.java:158]: to my mind, using ordinals is subtle and less clear logic

File: core/src/main/java/com/epam/esm/gift_certificate/dao/impl/SqlQueries.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
+package com.epam.esm.gift_certificate.dao.impl;
+
+public class SqlQueries {
+
+    /**
+     * Gift_certificate queries
+     */
+
+    static String SELECT_ALL_GIFT_CERTIFICATES_BY_TAG = "select g_id, g_name, g_description, g_price, g_duration" +
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/dao/impl/SqlQueries.java:9]: this approach is not flexible and maintenance becomes quite hard

File: core/src/main/java/com/epam/esm/gift_certificate/dao/impl/TagDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    @Override
+    public Optional<Tag> get(int id) throws NoSuchTagException {
+        try {
+            return Optional.ofNullable(jdbcOperations.queryForObject(
+                    SqlQueries.SELECT_TAG_BY_INDEX
+                    , this::mapTag
+                    , id));
+        } catch (IncorrectResultSizeDataAccessException exception) {
+            LOGGER.info(NO_SUCH_TAG_ID + id);
+            throw new NoSuchTagException(NO_SUCH_TAG_ID + id);
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/dao/impl/TagDaoImpl.java:57]: from the class name it clear enough that has happened, why we need message with same information?

File: core/src/main/java/com/epam/esm/gift_certificate/exception/NoSuchCertificateException.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+    public NoSuchCertificateException(String message) {
+        super(message);
+    }
+
+    public NoSuchCertificateException(String message, Throwable cause) {
+        super(message, cause);
+    }
+
+    public NoSuchCertificateException(Throwable cause) {
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/exception/NoSuchCertificateException.java:17]: do we use all three constructors in the app?

File: core/src/main/java/com/epam/esm/gift_certificate/model/dto/GiftCertificateDto.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+    private Double price;
+
+    private Integer duration;
+
+    private String createDate;
+
+    private String lastUpdateDate;
+
+    private List<Tag> tags;
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/model/dto/GiftCertificateDto.java:24]: it is better to use TagDto

File: core/src/main/java/com/epam/esm/gift_certificate/model/dto/GiftCertificateDto.java [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+    public GiftCertificateDto() {
+    }
+
+    public GiftCertificateDto(Integer id, String name, String description, Double price, Integer duration, String createDate, String lastUpdateDate, List<Tag> tags) {
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/model/dto/GiftCertificateDto.java:29]: do we use this constructor?

File: core/src/main/java/com/epam/esm/gift_certificate/model/entity/GiftCertificate.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    private Timestamp createDate;
+
+    private Timestamp lastUpdateDate;
+
+    private List<Tag> tags;
+
+    public GiftCertificate() {
+    }
+
+    public GiftCertificate(Integer id, String name, String description, Double price, Integer duration, Timestamp createDate, Timestamp lastUpdateDate, List<Tag> tags) {
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/model/entity/GiftCertificate.java:29]: do we use both constructors?

File: core/src/main/java/com/epam/esm/gift_certificate/service/api/GiftCertificateService.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+     * @return List of GiftCertificateDto
+     */
+    List<GiftCertificateDto> readAllGiftCertificates(ParamContext paramContext);
+
+    /**
+     * Retrieves list of GiftCertificates by tag name
+     * @param tag searching tag
+     * @return List of GiftCertificateDto
+     */
+    List<GiftCertificateDto> readAllGiftCertificatesByTag(String tag);
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/service/api/GiftCertificateService.java:29]: can we include tag to ParamContext?

File: core/src/main/java/com/epam/esm/gift_certificate/service/impl/GiftCertificateServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+        if (giftCertificateOptional.isPresent()) {
+            dto = mapToDto(giftCertificateOptional.get());
+            dto.setTags(tagDao.getTagsByGiftCertificateId(id));
+        }
+
+        return dto;
+    }
+
+    @Override
+    @SuppressWarnings("all")
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/service/impl/GiftCertificateServiceImpl.java:89]: why do we need this annotation?

File: core/src/main/java/com/epam/esm/gift_certificate/service/impl/GiftCertificateServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+    @Transactional
+    public GiftCertificateDto createGiftCertificate(GiftCertificateDto certificateDto)
+            throws NoSuchCertificateException, NoSuchTagException, TagCreationException {
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/service/impl/GiftCertificateServiceImpl.java:92]: why do we need checked exceptions?

File: core/src/main/java/com/epam/esm/gift_certificate/service/impl/GiftCertificateServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+
+        final List<Tag> tags = certificateDto.getTags();
+        final Integer newId = giftCertificateDao.create(mapToEntity(certificateDto));
+
+        if (tags != null) {
+            populateTagsCreate(newId, tags);
+        }
+
+        return readGiftCertificate(newId);
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/service/impl/GiftCertificateServiceImpl.java:101]: why do we need additional db call?

File: core/src/main/java/com/epam/esm/gift_certificate/service/impl/GiftCertificateServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    @SuppressWarnings("all")
+    private void populateTagsCreate(int giftCertificateId, List<Tag> insertedTags)
+            throws NoSuchTagException, TagCreationException {
+
+        List<Tag> allTags = tagDao.getAll();
+
+        for (Tag insertedTag : insertedTags) {
+            int newTagId = processNewTags(allTags, insertedTag);
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/service/impl/GiftCertificateServiceImpl.java:132]: `processNewTags ` not informative name

File: core/src/main/java/com/epam/esm/gift_certificate/service/impl/GiftCertificateServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,244 @@
...
+        List<Tag> actualTags = tagDao.getTagsByGiftCertificateId(giftCertificateId);
+        List<Tag> allTags = tagDao.getAll();
+
+        for (Tag insertedTag : insertedTags) {
+            int newTagId = processNewTags(allTags, insertedTag);
+
+            if (newTagId == 0) {
+                newTagId = allTags.stream()
+                        .filter(tag -> tag.getName().equals(insertedTag.getName())).findAny().get().getId();
+            }
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/service/impl/GiftCertificateServiceImpl.java:155]: code duplication

File: core/src/main/java/com/epam/esm/gift_certificate/service/impl/GiftCertificateServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,244 @@
...
+                , certificate.getName()
+                , certificate.getDescription()
+                , certificate.getPrice()
+                , certificate.getDuration()
+                , null
+                , null
+                , new ArrayList<>());
+    }
+
+    private String resolveQuery(ParamContext paramContext) {
```
Comment (vadimsemernik) [inline core/src/main/java/com/epam/esm/gift_certificate/service/impl/GiftCertificateServiceImpl.java:209]: this logic should be placed at dao level

File: core/src/test/java/com/epam/esm/gift_certificate/dao/GiftCertificateDaoImplTest.java [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+
+import javax.sql.DataSource;
+import java.util.ArrayList;
+
+
+public class GiftCertificateDaoImplTest {
+
+    private static GiftCertificate giftCertificate;
+
+    private ApplicationContext applicationContext;
```
Comment (vadimsemernik) [inline core/src/test/java/com/epam/esm/gift_certificate/dao/GiftCertificateDaoImplTest.java:18]: do we need applicationContext here?

File: gift-certificate-web-service/src/main/java/com/epam/esm/gift_certificate/config/ContextConfig.java [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+
+
+import javax.sql.DataSource;
+
+@Configuration
+@ComponentScan("com.epam.esm")
+@EnableWebMvc
+@EnableTransactionManagement
+@PropertySource("classpath:db.properties")
+public class ContextConfig implements WebMvcConfigurer {
```
Comment (vadimsemernik) [inline gift-certificate-web-service/src/main/java/com/epam/esm/gift_certificate/config/ContextConfig.java:29]: `ContextConfig ` is not informative name

File: gift-certificate-web-service/src/main/java/com/epam/esm/gift_certificate/controller/AppExceptionHandler.java [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+import java.util.Locale;
+
+@RestControllerAdvice
+@RequestMapping(consumes = "application/json", produces = "application/json")
```
Comment (vadimsemernik) [inline gift-certificate-web-service/src/main/java/com/epam/esm/gift_certificate/controller/AppExceptionHandler.java:18]: do we need `RequestMapping` here?

File: core/src/main/java/com/epam/esm/gift_certificate/model/dto/GiftCertificateDto.java (hunk 10 above)
Comment (pristavkin-ivan) [reply core/src/main/java/com/epam/esm/gift_certificate/model/dto/GiftCertificateDto.java]: no, but it's mandatory for json converter

**Pull request merged**